File: server/models/page_config.py
from pydantic import BaseModel, Field
from typing import List, Optional

class Node(BaseModel):
    id: str
    kind: str
    widget_type: Optional[str] = None
    props: Optional[dict] = {}
    style: Optional[dict] = {}
    advanced: Optional[dict] = {}
    children: Optional[List['Node']] = []

    class Config:
        schema_extra = {
            "example": {
                "id": "root",
                "kind": "root",
                "widget_type": None,
                "props": {},
                "style": {},
                "advanced": {},
                "children": []
            }
        }

class PageConfig(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    page_name: str
    sections: Optional[List[Node]] = []

    class Config:
        schema_extra = {
            "example": {
                "page_name": "home",
                "sections": [
                    {
                        "id": "root",
                        "kind": "root",
                        "children": []
                    }
                ]
            }
        }

File: server/models/PageConfig.js
const mongoose = require('mongoose');

// Node schema - matches the builder's node structure
const nodeSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true
  },
  kind: {
    type: String,
    enum: ['root', 'section', 'column', 'widget'],
    required: true
  },
  widgetType: String, // For widgets
  props: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  style: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  advanced: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  children: [{
    type: mongoose.Schema.Types.Mixed
  }]
}, { _id: false });

const pageConfigSchema = new mongoose.Schema({
  pageName: {
    type: String,
    enum: ['home', 'categories', 'products', 'about', 'contact'],
    required: true,
    unique: true
  },
  // Store the complete node tree structure from the builder
  sections: [{
    type: mongoose.Schema.Types.Mixed
  }],
  // Legacy: kept for backward compatibility
  meta: {
    title: String,
    description: String,
    keywords: String
  },
  isPublished: {
    type: Boolean,
    default: true
  },
  publishedAt: {
    type: Date,
    default: Date.now
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Update updatedAt before saving
pageConfigSchema.pre('save', function () {
  this.updatedAt = Date.now();
});

module.exports = mongoose.model('PageConfig', pageConfigSchema);

